ludwig/models/llm.py

```diff
@@ -40,9 +40,30 @@ def __init__(
         self.config_obj = config_obj
         self._random_seed = random_seed
 
-        self.model = AutoModelForCausalLM.from_pretrained(self.config_obj.model_name)
-        self.generation_config = GenerationConfig(**self.config_obj.generation_config.to_dict())
-        self.max_new_tokens = self.config_obj.generation_config.max_new_tokens
+        print("Loading large language model...")
+        self.model = AutoModelForCausalLM.from_pretrained(
+            self.config_obj.model_name,
+            low_cpu_mem_usage=True,
+            torch_dtype=torch.float16,
+            device_map="auto",
+            max_memory={i: "13GiB" for i in range(3)},
+        )
+        print("Done.")
+
+        # Used only for its metadata about the vocabulary
+        self.tokenizer = AutoTokenizer.from_pretrained(self.config_obj.model_name, use_fast=False)
+
+        # Determines the maximum length of the context (input + output tokens)
+        if hasattr(self.model.config, "max_sequence_length"):
+            self.context_len = self.model.config.max_sequence_length
+        elif hasattr(self.model.config, "max_position_embeddings"):
+            self.context_len = self.model.config.max_position_embeddings
+        else:
+            self.context_len = 2048
+
+        self.generation = GenerationConfig(**self.config_obj.generation.to_dict())
+        self.max_new_tokens = self.config_obj.generation.max_new_tokens
+        self.max_input_length = self.context_len - self.max_new_tokens - 8
```

Comment:
Why do we subtract 8? Is it because of padding alignment?

Reply:
Not sure, just copied this from @tgaddair's code. Maybe he can chime in?

Reply:
Upshot: it was in the fast chat code. Leaving as-is for now.